core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java

```diff
@@ -0,0 +1,132 @@
+     * @throws TimeoutException if timed out waiting for in-flight operations to finish
+     */
+    public void blockOperations(long timeout, TimeUnit timeUnit, Runnable onBlocked) throws InterruptedException, TimeoutException {
+        synchronized (mutex) {
+            if (blockOperationsInProgress) {
+                throw new IllegalStateException("cannot invoke blockOperations while it's running already");
+            }
+            blockOperationsInProgress = true;
+        }
+        try {
+            try (Releasable block = suspendableRefContainer.tryBlockAcquisition(timeout, timeUnit)) {
+                onBlocked.run();
+            }
+        } finally {
+            final List<ActionListener<Releasable>> queuedActions;
+            synchronized (mutex) {
+                assert blockOperationsInProgress;
+                blockOperationsInProgress = false;
+                queuedActions = delayedOperations;
+                delayedOperations = null;
+            }
+            if (queuedActions != null) {
+                for (ActionListener<Releasable> queuedAction : queuedActions) {
+                    acquire(queuedAction, null);
+                }
+            }
+        }
+    }
+
+    /**
+     * Acquires a lock whenever lock acquisition is not blocked. If the lock is directly available, the provided
+     * ActionListener will be called on the calling thread. During calls of {@link #blockOperations(long, TimeUnit, Runnable)}, lock
+     * acquisition can be delayed. The provided ActionListener will then be called using the provided executor once blockOperations
+     * terminates.
+     *
+     * @param onAcquired ActionListener that is invoked once acquisition is successful or failed
+     * @param executor executor to use for delayed call
+     */
+    public void acquire(ActionListener<Releasable> onAcquired, String executor) {
+        while (true) {
+            final Releasable releasable;
+            try {
+                releasable = suspendableRefContainer.tryAcquire();
+            } catch (InterruptedException e) {
+                onAcquired.onFailure(e);
+                return;
+            }
+            if (releasable != null) {
+                onAcquired.onResponse(releasable);
+                return;
+            } else {
+                // relocation hand-off in progress, store onRelocated callback to trigger after hand-off finished
+                synchronized (mutex) {
+                    if (blockOperationsInProgress) {
+                        // we know this will be released by call to releaseDelayedOperations
+                        if (delayedOperations == null) {
+                            delayedOperations = new ArrayList<>();
+                        }
+                        if (executor != null) {
+                            delayedOperations.add(new ThreadedActionListener(logger, threadPool, executor, onAcquired, true));
```

Comment:
forcing execution should be a parameter for now imo - I know we want/maybe/potentially change how we deal with replicas and queues, but for now I rather not change semantics and have primary ops non-forced and replicas ops forced.